### ISSUE TITLE
Caching with backward seeking is not optimized

### ISSUE THREAD

**Issue by hsn10** [mpv-player/mpv#10396]
**Describe the bug**
When seeking backward more cache is reloaded then it needs to be. Probably related to LRU caching method which is not optimal for dealing with streams.

**Test file**

Steps to reproduce the behaviour:

Test file is
_/ipfs/QmaWQgBv77QEPRgxzx3KrW3Y5oyzpbgGTpWdi9xh1PTRHS_   - its 178MB mp3 1h14m16s long

audio file from dead world radio episode 27 -> https://pin.ski/3pXsqht or https://crustwebsites.net/ipfs/Qmcp5Tu1Zg6sHj8JncRzXYFosMhuTtVn2f9WiUEzEHAJmP?filename=027

You need to access test file through some _IPFS gateway_ prepend gateway URL to IPFS address. Examples some gateways:
https://crustwebsites.net/ipfs/QmaWQgBv77QEPRgxzx3KrW3Y5oyzpbgGTpWdi9xh1PTRHS or https://gw.crustapps.net/ipfs/QmaWQgBv77QEPRgxzx3KrW3Y5oyzpbgGTpWdi9xh1PTRHS or
http://bafybeifuzqxrrvg7sm2feb2jlxlu4tat5gxcozrqdd3avalqwpeql7f4be.ipfs.cf-ipfs.com 

Its _not critical_ to get the same file used for test probably streaming any other mp3 with similar size and duration can be used instead.

**Problem**

1. we play the file from start and it will download and cache 43minutes. Indicated by line on screen. Cache size reported on console to be 150M.

1. Then we click and play at time 35 minutes. Cache will shift and will be from 16m to 1h14m. So far so good.

1. Then we click on 6 minute. Correct optimal behaviour will be to cache 10 minutes from 6m to 16m and free last 10 minutes because we already have rest downloaded.

1. This is not what is happening. mpv will load data at 6 minutes and moving forward and remove oldest blocks from cache which are at 16 minute. This will create moving hole and after we reach 16th minute when loading cache moving forward we would need to load data that were previously freed,

**Comment by ZaquL** [comment]
> 4\. would need to load data that were previously freed,

Mmm, and why is that surprising? Are you proposing not deletinb anything? Or are you proposing deleting from the end?

**Comment by hsn10** [comment]
We know what we need to download (cache size and player head position is known). Do intersection of "download plan" and "what is cached" - Do not release from cache any of these blocks

**Comment by alfonslm** [comment]
I still have this bad behaviour, it's way worse on 50gb movies since 2min is about 1gb. If the default behaviour is to evict old cache based on distance from the playhead instead of age this issue would be resolved, preferably with a default option to not evict cache based on size if the total cache is under the sum of forward and backwards and just use the size to prioritize if it's the forward or backward cache that should be evicted first to avoid unessesary eviction while jumping around.

Eg if we have 30 min forward cached as the forward max and 15min back max, jump forward eg 25min realize that it was way to long and jump back 10 min the current behavior would most likely have evicted the file based on the cache size even if the total cache size was fine (the units might not be the best one but it's pretty representative for the default behavior and close to what I have experiensed with less convinent numbers)

**Comment by kasper93** [comment]
@na-na-hi are you interested in this issue? I think solution could be to add an option for caching eviction strategy, unless we know that there is universally better strategy that we should always use.

**Comment by alfonslm** [comment]
The big question is probably CPU cycles but we can probably say that there's a way better default.

**Comment by na-na-hi** [comment]
> [@na-na-hi](https://github.com/na-na-hi) are you interested in this issue? I think solution could be to add an option for caching eviction strategy, unless we know that there is universally better strategy that we should always use.

I looked into this issue before. It's hard to say what is the best strategy here because seeking behavior and intent are not fixed. If I seek back to rewatch a few seconds of missing moment and then go back to the original position to continue playing, then the current LRU strategy is better than purging the further packet away from playback position. 

Also the current demux packet list is single linked list, and removing tail packets have bad performance.

**Comment by alfonslm** [comment]
Not sure if I speak for everyone but I think that basically everyone is saying that the backside of the current system is that you can quite easily cache half the file and then jump back and make the cache useless since it won't fill the gap ever.
Also, the current behaviour is probably worse in situations have long loading times so if you jump around less of the existing cache should be evicted while seeking in comparison to a fast pipeline where you probably just want to reduce peak load time on a drive.

**Comment by alfonslm** [comment]
> I looked into this issue before. It's hard to say what is the best strategy here because seeking behavior and intent are not fixed. If I seek back to rewatch a few seconds of missing moment and then go back to the original position to continue playing, then the current LRU strategy is better than purging the further packet away from playback position. 
> 
In this exempel you basically have two guesses, if you stopped paying attention eg because you got a call or if you want to know what someone said.
In one case the current behaviour risk having the entire buffer purged from jumping out of it and allowing to purge towards the playback risk having to partially reload a section of it. 
I will probably say that one of the behaviours is clearly worse when you load from a slow source

**Comment by kasper93** [comment]
> > [@na-na-hi](https://github.com/na-na-hi) are you interested in this issue? I think solution could be to add an option for caching eviction strategy, unless we know that there is universally better strategy that we should always use.
> 
> I looked into this issue before. It's hard to say what is the best strategy here because seeking behavior and intent are not fixed. If I seek back to rewatch a few seconds of missing moment and then go back to the original position to continue playing, then the current LRU strategy is better than purging the further packet away from playback position.

I agree that its not so obvious solution. I haven't think about this deeply, but like you said LRU sometimes works better than other times. 

> Also the current demux packet list is single linked list, and removing tail packets have bad performance.

That's not an issue, we can extend structure, but more importantly we can operate on ranges, which are finite number and select which one is preferably to be evicted.

Without much changes we could mark the 1st closest to play head forward cache range as protected and start evicting from the 2nd oldest (if first one was oldest). This will allow preserving and connecting the 1st forward range and then next one would be protected if not already evicted. This doesn't flip the caching upside down, only single range is affected, and fixes the "gap" issue.

**Comment by na-na-hi** [comment]
> Without much changes we could mark the 1st closest to play head forward cache range as protected and start evicting from the 2nd oldest (if first one was oldest). This will allow preserving and connecting the 1st forward range and then next one would be protected if not already evicted. This doesn't flip the caching upside down, only single range is affected, and fixes the "gap" issue.

If any cache range is protected it means the max cache size option is essentially not working as expected, as after a back seek it leaves no space to cache the current range. You either make cache grow larger than the limit or get constant buffering issues.

**Comment by kasper93** [comment]
> > Without much changes we could mark the 1st closest to play head forward cache range as protected and start evicting from the 2nd oldest (if first one was oldest). This will allow preserving and connecting the 1st forward range and then next one would be protected if not already evicted. This doesn't flip the caching upside down, only single range is affected, and fixes the "gap" issue.
> 
> If any cache range is protected it means the max cache size option is essentially not working as expected, as after a back seek it leaves no space to cache the current range. You either make cache grow larger than the limit or get constant buffering issues.

Of course if we don't have budget for cache or no back bytes to drain. That's not final solution, but the heuristic to "close the gap" is genuinely important fix here. We cannot predict where user will seek, but we can assume that in linear playback, trashing the cache just because there is small gap, is not correct behavior.
